DRA API: AdminAccess in DeviceRequestAllocationResult + DRAAdminAccess feature gate

pohly opened the pull request:
#### What type of PR is this?

/kind bug
/kind api-change

#### What this PR does / why we need it:

Drivers need to know that because admin access may also grant additional permissions. The allocator needs to ignore such results when determining which devices are considered as allocated. In both cases it is conceptually cleaner to not rely on the content of the ClaimSpec.

In preparation for beta in 1.32, a separate `DRAAdminAccess` feature gate gets added because the admin access feature needs to remain in alpha.

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
DRA: the `DeviceRequestAllocationResult` struct now has an "AdminAccess" field which should be used instead of the corresponding field in the `DeviceRequest` field when dealing with an allocation. If a device is only allocated for admin access, allocating it again for normal usage is now supported, as originally intended. To allow admin access, starting with 1.32 the `DRAAdminAccess` feature gate must be enabled.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4381
```

cc @klueska

pohly committed:
DRA API: AdminAccess in DeviceRequestAllocationResult

Drivers need to know that because admin access may also grant additional
permissions. The allocator needs to ignore such results when determining which
devices are considered as allocated.

In both cases it is conceptually cleaner to not rely on the content of the
ClaimSpec.

kannon92 commented:
/label api-review

pohly commented on staging/src/k8s.io/api/resource/v1alpha3/types.go:
```diff
@@ -785,6 +785,14 @@ type DeviceRequestAllocationResult struct {
 	Device string `json:"device" protobuf:"bytes,4,name=device"`
+
+	// AdminAccess indicates that administrative access to the device(s)
+	// was requested. A device where the only allocation is for
```
Does it make more sense after reading about the feature in the request?

https://github.com/kubernetes/kubernetes/blob/139cc3c659dd1624f7a4bbcc3b07fda79539677a/staging/src/k8s.io/api/resource/v1alpha3/types.go#L454-L462

kannon92 replied:
> A device where the only allocation is for administrative access is still available for other allocations.

How is that related to what you shared?

pohly replied:
The value in the status is a direct copy of the value in the spec.

I suppose that is what I need to call out explicitly. I thought it would be obvious :sweat_smile:

pohly replied:
> >    A device where the only allocation is for administrative access is still available for other allocations.
> How is that related to what you shared?

That follows from the definition of "admin access": admin access does not block normal usage.

pohly replied:
I replaced with "AdminAccess is a copy of the AdminAccess value in the request which caused this device to be allocated.".

What that means follows from the definition of that boolean in the request.

seans3 commented:
/triage accepted

kannon92 commented on staging/src/k8s.io/dynamic-resource-allocation/structured/allocator_test.go:
```diff
@@ -240,13 +240,17 @@ func slice(name string, nodeSelection any, pool, driver string, devices ...resou
 
-func deviceAllocationResult(request, driver, pool, device string) resourceapi.DeviceRequestAllocationResult {
-	return resourceapi.DeviceRequestAllocationResult{
+func deviceAllocationResult(request, driver, pool, device string, adminAccess ...bool) resourceapi.DeviceRequestAllocationResult {
```
Is there any reason why we are grabbing only the zeroth entry of this array?

Should we have a test case where the adminAccess differs in the array?

pohly replied:
This is a quick-and-dirty way of having an optional parameter. There's never a situation where `adminAccess` has more than one entry.

pohly replied:
Too dirty! I replaced this with a simple boolean argument.

haircommander commented:
/assign @kannon92

pohly commented:
/assign @thockin 

The first of the 1.32 API changes. I'm unsure whether I should treat this as an API break or not.

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -779,6 +779,12 @@ type DeviceRequestAllocationResult struct {
+	// request which caused this device to be allocated.
+	//
+	// +optional
+	AdminAccess bool
```
@klueska @byako: if you intend to support 1.31 and 1.32, then there is a slight problem: your driver cannot distinguish between "false" (no admin access with 1.32) and "not set" (behavior in 1.31, where only the spec had the original field).

One way to ensure interoperability would be to turn this into a pointer and explicitly document that "if this is not set, the corresponding field in the request in the spec needs to be checked".

This might be cleaner.

pohly replied:
I think it is, so I changed it: https://github.com/kubernetes/kubernetes/compare/88f8eb35dace02314007ee33e4cd73130f75e5ab..829071124985045dbeee84b4b03feac89b024968

byako replied:
Why would we need to distinguish? As long as we know when it is True - that's what matters, isn't it?

pohly replied:
Making it a pointer enables clients to detect "old" allocations where admin access was defined only in the spec.

If a client doesn't care to support 1.31, it can always use `ptr.Deref(, false)`.

pohly replied:
Upgrades are allowed to break when alpha features are involved, but that doesn't mean that they are automatically unsupported or discouraged. If it's possible to support them, then it's better to do so.

The main practical purpose is to enable @klueska to support admin access in a driver release that then works with 1.31 and 1.32.

Doing it "cleanly" according to API guidelines is less important in practice. We could pretend that 1.31 doesn't exist and just use `AdminAccess bool` like I did initially. It certainly was easier (no need for validation, no code changes in various places to set the field in new allocations).

pohly replied:
I changed the wording again a bit by adding "... if the field is unset.".

@kannon92, @byako, @klueska: is this good now for API review? Someone please LGTM.

kannon92 commented:
/retest

kannon92 commented on pkg/apis/resource/validation/validation_resourceclaim_test.go:
```diff
@@ -416,10 +420,31 @@ func TestValidateClaimStatusUpdate(t *testing.T) {
 							Driver:  goodName,
 							Pool:    goodName,
 							Device:  goodName,
+							// AdminAccess: ptr.To(false),
```
If you want to leave this in, I'd probably comment on it.

Whenever I see comments I think it was an accident to be left behind. In this case, I guess you wanted to add a comment to make it clear that AdminAccess is not specified.

kannon92 replied:
TBH the test name is clear enough without the comment IMO.

pohly replied:
Changed to:
```
AdminAccess: nil, // Intentionally not set.
```

I want to call out the difference compared to the other case.

kannon92 commented on staging/src/k8s.io/dynamic-resource-allocation/structured/allocator_test.go:
```diff
@@ -240,13 +240,15 @@ func slice(name string, nodeSelection any, pool, driver string, devices ...resou
 
-func deviceAllocationResult(request, driver, pool, device string) resourceapi.DeviceRequestAllocationResult {
-	return resourceapi.DeviceRequestAllocationResult{
+func deviceAllocationResult(request, driver, pool, device string, adminAccess bool) resourceapi.DeviceRequestAllocationResult {
```
To make this a bit easier in the future, maybe you could define a constant for adminAccess so that it is clear what these parameter is during the test code.

Usually for API fields, we discourage bools as they are not very that descriptive but I think this is probably fine as a bool. But we sho

pohly replied:
Should what?

kannon92 commented on staging/src/k8s.io/dynamic-resource-allocation/structured/allocator_test.go:
```diff
@@ -716,6 +718,42 @@ func TestAllocator(t *testing.T) {
+			node:    node(node1, region1),
+
+			expectResults: []any{
+				allocationResult(localNodeSelector(node1), deviceAllocationResult(req0, driverA, pool1, device1, false)),
```
This seems counter intuitive to me. Why would you not want to report that this the allocation has admin access?

pohly replied:
The *new* request in `claimsToAllocate` isn't using admin access, so the expected result doesn't either.

The *old* allocations in `allocatedClaims` use it, therefore they don't block using the device for a normal claim.

kannon92 commented:
/lgtm

I think its ready for api-review.

pohly commented:
/assign @thockin 

For API review. This includes a bug fix, therefore I'd like to get it merged separately for the release note entry.

pohly committed:
DRA admin access: add feature gate

The new DRAAdminAccess feature gate has the following effects:
- If disabled in the apiserver, the spec.devices.requests[*].adminAccess
  field gets cleared. Same in the status. In both cases the scenario
  that it was already set and a claim or claim template get updated
  is special: in those cases, the field is not cleared.

  Also, allocating a claim with admin access is allowed regardless of the
  feature gate and the field is not cleared. In practice, the scheduler
  will not do that.
- If disabled in the resource claim controller, creating ResourceClaims
  with the field set gets rejected. This prevents running workloads
  which depend on admin access.
- If disabled in the scheduler, claims with admin access don't get
  allocated. The effect is the same.

The alternative would have been to ignore the fields in claim controller and
scheduler. This is bad because a monitoring workload then runs, blocking
resources that probably were meant for production workloads.

pohly commented:
/wg device-management

klueska commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -279,9 +278,14 @@ func validateResourceClaimStatusUpdate(status, oldStatus *resource.ResourceClaim
 
-	// Updates to a populated status.Allocation are not allowed
+	// Updates to a populated status.Allocation are not allowed.
+	// Unmodified fields don't need to be validated again and,
+	// in this particular case, must not be validated again because
+	// validation for new results is tighter than it was before.
 	if oldStatus.Allocation != nil && status.Allocation != nil {
 		allErrs = append(allErrs, apimachineryvalidation.ValidateImmutableField(status.Allocation, oldStatus.Allocation, fldPath.Child("allocation"))...)
+	} else if status.Allocation != nil {
+		allErrs = append(allErrs, validateAllocationResult(status.Allocation, fldPath.Child("allocation"), requestNames)...)
```
Is this related to this change? I can see why its needed, but it seems unrelated to providing AdminAccess field as part of the DeviceRequestAllocationResult. Does it at least deserve its own commit?

pohly replied:
We only needed it because of the special logic that is required for fields that might be invalid on an update, so it belongs in the same commit? Otherwise we probably would validate again.

klueska commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -361,6 +364,9 @@ func validateDeviceRequestAllocationResult(result resource.DeviceRequestAllocati
 	allErrs = append(allErrs, validateDriverName(result.Driver, fldPath.Child("driver"))...)
 	allErrs = append(allErrs, validatePoolName(result.Pool, fldPath.Child("pool"))...)
 	allErrs = append(allErrs, validateDeviceName(result.Device, fldPath.Child("device"))...)
+	if result.AdminAccess == nil {
+		allErrs = append(allErrs, field.Required(fldPath.Child("adminAccess"), ""))
+	}
```
Should we create a `validateAdminAccess()` helper for symmetry with the rest of these calls?

klueska reviewed (commented):
Minor comment, otherwise looks good to me. I'll leave it up to API review if it makes sense to have AdminAccess be a pointer or just continue to be a bool and *not* support DRA drivers that made use of it in 1.31. There are no known drivers that do, so it may simplify things.

pohly commented:
> I'll leave it up to API review if it makes sense to have AdminAccess be a pointer or just continue to be a bool and not support DRA drivers that made use of it in 1.31.

Also worth considering in this context: with the upcoming feature gate for admin access, "feature gate off" is another situation in 1.32 where the field is "unset" instead of "set to false", but I am unsure whether it makes any difference because if the request has the field set, I intend to allow setting it also in the status because that's then internally consistent.

I'll add the feature gate check as a second commit here. It's probably better to review both together.

pohly commented:
This conflicts with https://github.com/kubernetes/kubernetes/pull/128003. I've rebased already on top of that. Should I push or wait?

If I push now, it'll add other, unrelated commits. The advantage would be that actual code reviews can then look at the code as it will go into master (i.e. review my merge conflict resolution).

pohly commented:
As I need to fix one E2E test, I am going to do that with my rebased branch and push that.

pohly commented:
/retest

cgroupv1 tests were failing because of a borked update in master which got reverted.

johnbelamaric commented on pkg/apis/resource/types.go:
```diff
@@ -442,8 +442,12 @@ type DeviceRequest struct {
 	// +optional
 	// +default=false
+	// +featureGate=DRAAdminAccess
 	AdminAccess bool
```
How does it work for this to be optional but not a pointer?

pohly replied:
We have done a few times in the API, by treating "false" the same as "not set". That's okay as long as we have no-one who needs to distinguish between the two, which is the case here.

pohly replied:
I've changed it because it's consistent with usual practice and thus easier to review. However, I don't think it makes a difference and is worse in one regard.

@LionelJouin, @aojea: you may have to reconsider this in the device status PR because you also have a feature gate.

Here, it doesn't make a difference for users because for them `adminAccess: false` and `adminAccess: null` are identical (= admin access off). They are probably not going to set it to false explicitly. It *would* make a difference if the field was mutable because then a user could set `adminAccess: false` and then because of the "allow setting disabled field if in use" rule change it later when the feature gate is turned off. But the claim spec and allocation status are immutable, so this special case does not apply here.

It is (slightly) worse because in the allocation status, "not set" in the first commit meant "look it up in the spec" to support upgrades from 1.31 where that status field didn't exist yet. Now "not set" means "off", the same as "false". This is entirely theoretical because there is no DRA driver which supported admin access with 1.31 and therefore probably no claim that could trigger this corner case, so it's okay to just add the status field and treat it as if we were adding this feature anew.

@thockin: I had to rebase, so those changes are in a new commit at the end in case that you want to do an incremental review.

johnbelamaric commented on pkg/registry/resource/resourceclaimtemplate/strategy.go:
```diff
@@ -92,3 +98,20 @@ func toSelectableFields(template *resource.ResourceClaimTemplate) fields.Set {
+
+	for i := range newClaimTemplate.Spec.Spec.Devices.Requests {
+		if oldClaimTemplate == nil || i >= len(oldClaimTemplate.Spec.Spec.Devices.Requests) || !oldClaimTemplate.Spec.Spec.Devices.Requests[i].AdminAccess {
+			newClaimTemplate.Spec.Spec.Devices.Requests[i].AdminAccess = false
```
I guess this sort of answers my question above - we treat "false" as dropped?

johnbelamaric reviewed (commented):
got started, will finish up by Monday at the latest

pohly commented:
/retest

thockin commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -331,6 +336,9 @@ func validateDeviceRequestAllocationResult(result resource.DeviceRequestAllocati
 	allErrs = append(allErrs, validateDriverName(result.Driver, fldPath.Child("driver"))...)
 	allErrs = append(allErrs, validatePoolName(result.Pool, fldPath.Child("pool"))...)
 	allErrs = append(allErrs, validateDeviceName(result.Device, fldPath.Child("device"))...)
+	if result.AdminAccess == nil && utilfeature.DefaultFeatureGate.Enabled(features.DRAAdminAccess) {
```
As noted elsewhere, you can't have really required fields behind gates.  One of the rules of API compatibility is that clients which used to work still work.  If the main API goes beta, clients will use that without this field.  Those clients can't be expected to magically learn about this field, unless there is a default value (e.g. defaults.go) which I do not see in this PR.

pohly replied:
Removed.

This was meant to ensure that if the feature gate is on, clients which have not been updated or have the feature disabled don't allocate a claim "incorrectly", but I see how that doesn't follow normal practices and also affects claims not using the feature.

thockin commented on pkg/registry/resource/resourceclaim/strategy.go:
```diff
@@ -172,4 +174,38 @@ func toSelectableFields(claim *resource.ResourceClaim) fields.Set {
+}
+
+func dropDisabledDRAAdminAccessFields(newClaim, oldClaim *resource.ResourceClaim) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.DRAAdminAccess) {
```
Usually this is "if gate-enabled || alread-in-use" - is it intentional to drop the 2nd part of that?

pohly replied:
I am using that pattern now. Instead of trying to be (too) smart and checking if the feature is set in the old request, I'm now just checking whether it was used anywhere in the old claim.

thockin commented on pkg/registry/resource/resourceclaim/strategy.go:
```diff
@@ -172,4 +174,38 @@ func toSelectableFields(claim *resource.ResourceClaim) fields.Set {
+		return
+	}
+
+	for i := range newClaim.Spec.Devices.Requests {
```
Are you 100% sure that you can associate old and new requests based on index?

pohly replied:
When assuming that specs are immutable, yes. But this code shouldn't make that assumption.

Anyway, I simplified it ("in use anywhere in the claim").

thockin commented on pkg/registry/resource/resourceclaim/strategy.go:
```diff
@@ -172,4 +174,38 @@ func toSelectableFields(claim *resource.ResourceClaim) fields.Set {
+	if newClaim.Status.Allocation == nil {
+		return
+	}
+	for i := range newClaim.Status.Allocation.Devices.Results {
```
Can this possibly be simplified?

pohly committed:
DRA: treat AdminAccess as a new feature gated field

Using the "normal" logic for a feature gated field simplifies the
implementation of the feature gate.

There is one (entirely theoretic!) problem with updating from 1.31: if a claim
was allocated in 1.31 with admin access, the status field was not set because
it didn't exist yet. If a driver now follows the current definition of "unset =
off", then it will not grant admin access even though it should. This is
theoretic because drivers are starting to support admin access with 1.32, so
there shouldn't be any claim where this problem could occur.

pohly commented on staging/src/k8s.io/api/resource/v1alpha3/types.go:
```diff
@@ -756,6 +760,22 @@ type DeviceRequestAllocationResult struct {
+	// This is an alpha field and requires enabling the DRAAdminAccess
+	// feature gate.
+	//
+	// +required
```
Ack.

thockin commented on pkg/registry/resource/resourceclaim/strategy.go:
```diff
@@ -172,4 +174,49 @@ func toSelectableFields(claim *resource.ResourceClaim) fields.Set {
+		return
+	}
+
+	for i := range newClaim.Spec.Devices.Requests {
```
FWIW, I think this could just be:

```
if utilfeature.DefaultFeatureGate.Enabled(features.DRAAdminAccess) {
    return
}
if draAdminAccessFeatureInUse(oldClaim) {
    return
}

for i := range newClaim.Spec.Devices.Requests {
    newClaim.Spec.Devices.Requests[i].AdminAccess = nil
}
if newClaim.Status.Allocation == nil {
    return
}
for i := range newClaim.Status.Allocation.Devices.Results {
    newClaim.Status.Allocation.Devices.Results[i].AdminAccess = nil
}
```

Couldn't it?  :)

pohly replied:
Yes. My thinking was that if there is nothing to drop, then I don't need to call `draAdminAccessFeatureInUse`, which should be the most common case. But I am iterating over both slices either way, so that doesn't really make sense.

But it's good that this is merged because it unblocks next steps. I'll fix this in a [follow-up PR](https://github.com/kubernetes/kubernetes/pull/128433).

thockin reviewed (commented):
Thanks!

/lgtm
/approve

pacoxu commented:
/test pull-kubernetes-node-e2e-containerd
